File: sdks/python/src/ai_control_sdk/runtime/routing.py
"""Deterministic current Runtime Snapshot route selection."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

from pydantic import BaseModel, ConfigDict, Field

from ..errors import AiControlSdkError
from .contracts import (
    RuntimeCallSourceMatch,
    RuntimeOverrideMatch,
    RuntimeRoute,
    RuntimeRouteTarget,
    RuntimeRoutingPlan,
    RuntimeRoutingRule,
    RuntimeScheduleMatch,
    RuntimeSnapshot,
    RuntimeUserMatch,
    RuntimeUserPropertyMatch,
    parse_utc,
)


class RuntimeRouteContext(BaseModel):
    model_config = ConfigDict(frozen=True, strict=True)

    user_id: str | None = None
    user_properties: dict[str, Any] = Field(default_factory=dict)
    call_source: str | None = None
    selection_key: str | None = None


class RuntimeRouteSelection(BaseModel):
    model_config = ConfigDict(frozen=True, strict=True)

    virtual_model: str
    virtual_model_id: str
    configuration_version: int
    configuration_etag: str
    route_tag: str
    rule_id: str | None
    primary: RuntimeRouteTarget
    fallbacks: tuple[RuntimeRouteTarget, ...]


def _minute(value: str) -> int:
    return int(value[:2]) * 60 + int(value[3:])


def _property_matches(rule: RuntimeUserPropertyMatch, context: RuntimeRouteContext) -> bool:
    condition = rule.user_property
    current = context.user_properties.get(condition.key)
    if condition.operator == "is_set":
        return current is not None
    if condition.operator == "is_not_set":
        return current is None
    if condition.operator == "equals":
        return current == condition.value
    if condition.operator == "not_equals":
        return current != condition.value
    if condition.operator == "starts_with":
        return isinstance(current, str) and current.startswith(str(condition.value))
    if isinstance(current, str):
        return str(condition.value) in current
    return isinstance(current, list) and condition.value in current


def _schedule_matches(rule: RuntimeScheduleMatch, instant: datetime, timezone: str) -> bool:
    try:
        local = instant.astimezone(ZoneInfo(timezone))
    except ZoneInfoNotFoundError as error:
        raise AiControlSdkError(
            "SDK_RUNTIME_TIMEZONE_INVALID", "Cannot evaluate route timezone."
        ) from error
    schedule = rule.schedule
    start, end = _minute(schedule.from_), _minute(schedule.to)
    weekday, minute = local.isoweekday(), local.hour * 60 + local.minute
    if start < end:
        return weekday in schedule.days and start <= minute < end
    if start > end:
        previous = 7 if weekday == 1 else weekday - 1
        return (weekday in schedule.days and minute >= start) or (
            previous in schedule.days and minute < end
        )
    return False


def _matches(
    rule: RuntimeRoutingRule,
    instant: datetime,
    plan: RuntimeRoutingPlan,
    context: RuntimeRouteContext,
) -> bool:
    if isinstance(rule.match, RuntimeOverrideMatch):
        return rule.expires_at is not None and parse_utc(rule.expires_at) > instant
    if isinstance(rule.match, RuntimeUserMatch):
        return context.user_id in rule.match.user.ids
    if isinstance(rule.match, RuntimeUserPropertyMatch):
        return _property_matches(rule.match, context)
    if isinstance(rule.match, RuntimeCallSourceMatch):
        return context.call_source == rule.match.call_source.value
    if isinstance(rule.match, RuntimeScheduleMatch):
        return _schedule_matches(rule.match, instant, plan.timezone)
    return False


def _selection(
    virtual_model: str,
    plan: RuntimeRoutingPlan,
    route: RuntimeRoute,
    rule_id: str | None,
    context: RuntimeRouteContext,
) -> RuntimeRouteSelection:
    targets = _ordered_targets(route, context.selection_key or context.user_id)
    return RuntimeRouteSelection(
        virtual_model=virtual_model,
        virtual_model_id=plan.virtual_model_id,
        configuration_version=plan.configuration_version,
        configuration_etag=plan.configuration_etag,
        route_tag=route.route_tag,
        rule_id=rule_id,
        primary=targets[0],
        fallbacks=tuple(targets[1:]),
    )


def _deterministic_fraction(value: str) -> float:
    hash_value = 2_166_136_261
    for byte in value.encode("utf-8"):
        hash_value ^= byte
        hash_value = (hash_value * 16_777_619) & 0xFFFFFFFF
    return hash_value / 4_294_967_296


def _ordered_targets(
    route: RuntimeRoute, selection_key: str | None
) -> tuple[RuntimeRouteTarget, ...]:
    targets = tuple(route.targets)
    if route.selection_mode != "weighted" or selection_key is None:
        return targets
    point = _deterministic_fraction(f"{route.route_tag}:{selection_key}") * sum(
        target.weight for target in targets
    )
    cumulative = 0.0
    selected = targets[0]
    for target in targets:
        cumulative += target.weight
        if point < cumulative:
            selected = target
            break
    return (selected, *(target for target in targets if target.model_id != selected.model_id))


def resolve_runtime_route(
    snapshot: RuntimeSnapshot,
    virtual_model: str,
    instant: datetime,
    context: RuntimeRouteContext | None = None,
) -> RuntimeRouteSelection:
    plan = snapshot.routing.get(virtual_model)
    if plan is None:
        raise AiControlSdkError(
            "SDK_RUNTIME_ROUTE_NOT_FOUND",
            f"No active route is published for virtual model {virtual_model}.",
        )
    route_context = context or RuntimeRouteContext()
    matched = [rule for rule in plan.rules if _matches(rule, instant, plan, route_context)]
    if not matched:
        return _selection(virtual_model, plan, plan.default, None, route_context)
    highest = max(rule.priority for rule in matched)
    winners = [rule for rule in matched if rule.priority == highest]
    if len(winners) != 1:
        raise AiControlSdkError(
            "SDK_RUNTIME_ROUTE_CONFLICT",
            f"Multiple route rules share the winning priority for {virtual_model}.",
        )
    winner = winners[0]
    return _selection(virtual_model, plan, winner.route, winner.id, route_context)

File: sdks/node/src/runtime/routing.ts
import type { RuntimeRoute, RuntimeRouteTarget, RuntimeSnapshot } from "@tokenpilot/contracts";

import { AiControlSdkError } from "../errors.js";

export interface RuntimeRouteSelection {
  readonly virtualModel: string;
  readonly virtualModelId: string;
  readonly configurationVersion: number;
  readonly configurationEtag: string;
  readonly routeTag: string;
  readonly ruleId: string | null;
  readonly primary: RuntimeRouteTarget;
  readonly fallbacks: readonly RuntimeRouteTarget[];
}

export interface RuntimeRouteContext {
  readonly userId?: string;
  readonly userProperties?: Readonly<Record<string, unknown>>;
  readonly callSource?: string;
  readonly selectionKey?: string;
}

const weekdayNumbers: Readonly<Record<string, number>> = Object.freeze({
  Mon: 1,
  Tue: 2,
  Wed: 3,
  Thu: 4,
  Fri: 5,
  Sat: 6,
  Sun: 7,
});

function localized(now: Date, timezone: string): { weekday: number; minute: number } {
  const parts = new Intl.DateTimeFormat("en-US", {
    timeZone: timezone,
    weekday: "short",
    hour: "2-digit",
    minute: "2-digit",
    hourCycle: "h23",
  }).formatToParts(now);
  const value = Object.fromEntries(parts.map((part) => [part.type, part.value]));
  const weekday = weekdayNumbers[value.weekday ?? ""];
  if (weekday === undefined) {
    throw new AiControlSdkError("SDK_RUNTIME_TIMEZONE_INVALID", "Cannot evaluate route timezone.");
  }
  return { weekday, minute: Number(value.hour) * 60 + Number(value.minute) };
}

function clockMinute(value: string): number {
  return Number(value.slice(0, 2)) * 60 + Number(value.slice(3, 5));
}

function previousWeekday(value: number): number {
  return value === 1 ? 7 : value - 1;
}

function ruleMatches(
  rule: RuntimeSnapshot["routing"][string]["rules"][number],
  local: { weekday: number; minute: number },
  now: Date,
  context: RuntimeRouteContext,
): boolean {
  if ("override_active" in rule.match) {
    return rule.expires_at !== undefined && new Date(rule.expires_at).getTime() > now.getTime();
  }
  if ("user" in rule.match) return rule.match.user.ids.includes(context.userId ?? "");
  if ("user_property" in rule.match) {
    const condition = rule.match.user_property;
    const value = context.userProperties?.[condition.key];
    if (condition.operator === "is_set") return value !== undefined && value !== null;
    if (condition.operator === "is_not_set") return value === undefined || value === null;
    if (condition.operator === "equals") return value === condition.value;
    if (condition.operator === "not_equals") return value !== condition.value;
    if (condition.operator === "starts_with") {
      return typeof value === "string" && value.startsWith(String(condition.value));
    }
    if (typeof value === "string") return value.includes(String(condition.value));
    return Array.isArray(value) && value.includes(condition.value);
  }
  if ("call_source" in rule.match) return context.callSource === rule.match.call_source.value;
  if (!("schedule" in rule.match)) return false;
  const schedule = rule.match.schedule;
  const start = clockMinute(schedule.from);
  const end = clockMinute(schedule.to);
  if (start < end) {
    return schedule.days.includes(local.weekday) && local.minute >= start && local.minute < end;
  }
  if (start > end) {
    return (
      (schedule.days.includes(local.weekday) && local.minute >= start) ||
      (schedule.days.includes(previousWeekday(local.weekday)) && local.minute < end)
    );
  }
  return false;
}

function selection(
  virtualModel: string,
  plan: RuntimeSnapshot["routing"][string],
  route: RuntimeRoute,
  ruleId: string | null,
  context: RuntimeRouteContext,
): RuntimeRouteSelection {
  const targets = orderedTargets(route, context.selectionKey ?? context.userId);
  const [primary, ...fallbacks] = targets;
  if (primary === undefined) {
    throw new AiControlSdkError("SDK_RUNTIME_ROUTE_EMPTY", "Selected runtime route has no target.");
  }
  return Object.freeze({
    virtualModel,
    virtualModelId: plan.virtual_model_id,
    configurationVersion: plan.configuration_version,
    configurationEtag: plan.configuration_etag,
    routeTag: route.route_tag,
    ruleId,
    primary: Object.freeze({ ...primary }),
    fallbacks: Object.freeze(fallbacks.map((target) => Object.freeze({ ...target }))),
  });
}

function deterministicFraction(value: string): number {
  let hash = 2_166_136_261;
  for (const byte of new TextEncoder().encode(value)) {
    hash ^= byte;
    hash = Math.imul(hash, 16_777_619);
  }
  return (hash >>> 0) / 4_294_967_296;
}

function orderedTargets(
  route: RuntimeRoute,
  selectionKey: string | undefined,
): readonly RuntimeRouteTarget[] {
  if (route.selection_mode !== "weighted" || selectionKey === undefined) return route.targets;
  const total = route.targets.reduce((sum, target) => sum + target.weight, 0);
  const point = deterministicFraction(`${route.route_tag}:${selectionKey}`) * total;
  let cumulative = 0;
  let selected = route.targets[0]!;
  for (const target of route.targets) {
    cumulative += target.weight;
    if (point < cumulative) {
      selected = target;
      break;
    }
  }
  return [selected, ...route.targets.filter((target) => target.model_id !== selected.model_id)];
}

export function resolveRuntimeRoute(
  snapshot: RuntimeSnapshot,
  virtualModel: string,
  now: Date,
  context: RuntimeRouteContext = {},
): RuntimeRouteSelection {
  const plan = snapshot.routing[virtualModel];
  if (plan === undefined) {
    throw new AiControlSdkError(
      "SDK_RUNTIME_ROUTE_NOT_FOUND",
      `No active route is published for virtual model ${virtualModel}.`,
    );
  }
  const local = localized(now, plan.timezone);
  const matched = plan.rules.filter((rule) => ruleMatches(rule, local, now, context));
  if (matched.length === 0) return selection(virtualModel, plan, plan.default, null, context);
  const highest = Math.max(...matched.map((rule) => rule.priority));
  const winners = matched.filter((rule) => rule.priority === highest);
  if (winners.length !== 1) {
    throw new AiControlSdkError(
      "SDK_RUNTIME_ROUTE_CONFLICT",
      `Multiple route rules share the winning priority for ${virtualModel}.`,
    );
  }
  const winner = winners[0]!;
  return selection(virtualModel, plan, winner.route, winner.id, context);
}
